kernel/src/scan/mod.rs

```diff
@@ -92,25 +96,166 @@ impl ScanBuilder {
         let logical_schema = self
             .schema
             .unwrap_or_else(|| self.snapshot.schema().clone().into());
-        let (all_fields, read_fields, have_partition_cols) = get_state_info(
+        let state_info = get_state_info(
             logical_schema.as_ref(),
             &self.snapshot.metadata().partition_columns,
-            self.snapshot.column_mapping_mode,
         )?;
-        let physical_schema = Arc::new(StructType::new(read_fields));
 
-        // important! before a read/write to the table we must check it is supported
-        self.snapshot.protocol().ensure_read_supported()?;
+        let physical_predicate = match self.predicate {
+            Some(predicate) => Self::build_physical_predicate(predicate, &logical_schema)?,
+            None => PhysicalPredicate::None,
+        };
 
         Ok(Scan {
             snapshot: self.snapshot,
             logical_schema,
-            physical_schema,
-            predicate: self.predicate,
-            all_fields,
-            have_partition_cols,
+            physical_schema: Arc::new(StructType::new(state_info.read_fields)),
+            physical_predicate,
+            all_fields: state_info.all_fields,
+            have_partition_cols: state_info.have_partition_cols,
         })
     }
+
+    // If we have a predicate, verify the columns it references and apply column mapping. First, get
+    // the set of references; use that to filter the schema to only the columns of interest (and
+    // verify that all referenced columns exist); then use the resulting logical/physical mappings
+    // to rewrite the expression with physical column names.
+    //
+    // NOTE: It is possible the predicate resolves to FALSE even ignoring column references,
+    // e.g. `col > 10 AND FALSE`. Such predicates can statically skip the whole query.
+    fn build_physical_predicate(
+        predicate: ExpressionRef,
+        logical_schema: &Schema,
+    ) -> DeltaResult<PhysicalPredicate> {
+        if can_statically_skip_all_files(&predicate) {
+            return Ok(PhysicalPredicate::StaticSkipAll);
+        }
+        let mut get_referenced_fields = GetReferencedFields {
```

Comment:
Not really... it would just factor them out to the `new` function. And since this struct is only constructed once -- here -- that would be a net increase in code. As soon as a second call site appears, `new` makes a _lot_ of sense tho!